File: src/main/java/com/leijendary/spring/microservicetemplate/document/LocalizedDocument.java
package com.leijendary.spring.microservicetemplate.document;

import lombok.Data;
import org.springframework.data.elasticsearch.annotations.Field;

import java.util.HashSet;
import java.util.Set;

import static com.leijendary.spring.microservicetemplate.util.RequestContext.getLanguage;
import static java.util.Comparator.comparingInt;
import static java.util.stream.Collectors.toList;
import static org.springframework.data.elasticsearch.annotations.FieldType.Nested;

@Data
public abstract class LocalizedDocument<T extends LocaleDocument> {

    @Field(type = Nested, includeInParent = true)
    private Set<T> translations = new HashSet<>();

    public T getTranslation() {
        final var language = getLanguage();
        final var sorted = translations
                .stream()
                .sorted(comparingInt(LocaleDocument::getOrdinal))
                .collect(toList());

        return sorted
                .stream()
                .filter(t -> t.getLanguage().equals(language))
                .findFirst()
                .orElse(sorted.iterator().next());
    }
}

File: src/main/kotlin/com/leijendary/spring/template/core/document/LocalizedDocument.kt
package com.leijendary.spring.template.core.document

import com.leijendary.spring.template.core.util.RequestContext.language
import org.springframework.data.elasticsearch.annotations.Field
import org.springframework.data.elasticsearch.annotations.FieldType.Nested

open class LocalizedDocument<T : LocaleDocument> {
    @Field(type = Nested, includeInParent = true)
    var translations = HashSet<T>()

    val translation: T
        get() {
            val language = language
            val sorted = translations.sortedBy { it.ordinal }

            return sorted.firstOrNull { it.language == language } ?: sorted.first()
        }
}
